src/JustSaying/Messaging/Channels/Receive/MessageReceiveBuffer.cs

```diff
@@ -70,6 +73,8 @@ public async Task RunAsync(CancellationToken stoppingToken)
             {
                 stoppingToken.ThrowIfCancellationRequested();
 
+                CheckMessageReceiveControllerStatus(stoppingToken);
```

Comment:
Hi, thanks for the swift feedback on a draft! Yes it would be simpler to use the existing while loop but it might be better to avoid a busy wait if possible to reduce CPU usage on the offline version of the service. However perhaps it can be used in relation with this comment - if an `is` check fails then fall back to the busy wait method?
https://github.com/justeattakeaway/JustSaying/pull/1164#discussion_r1186173424

Using an internal enum instead of a boolean could be nicer 👍

Reply:
Would using a `Task.Delay(100)` on NotReceiving avoid any CPU thrashing on the loop, while also avoiding the complexity of the PauseThreads technique?

Reply:
We can test a larger delay than what was tried previously, and maybe the value can be made configurable. If the simpler method is preferrable and it doesn't increase CPU too much I can open a separate PR

Reply:
Ah interesting, was there a previous attempt? Whereabouts?

A sleep of even 10ms would be millions of cycles skipped, so I wouldn't worry too much about CPU usage - it will be effectively zero.

Reply:
In https://github.com/justeattakeaway/JustSaying/pull/1164/commits/17e3cc6b93a1b0884aae4aa5b01125453e98fa82. Not quite the same as you described previously as it uses another while loop, but the same concept.

I have changed it locally so it just `continue`s in the main while loop as suggested, and used 100ms. CPU usage is a bit lower but still not as low as the thread pausing method, roughly double. This is a graph of the service process on a test instance:

![image](https://github.com/justeattakeaway/JustSaying/assets/92371891/bd724949-b395-4879-a020-13e750c39f62)

Blue being the latest changes (the spike is deployment), green the previous attempt with 1ms and yellow with the threads paused. I understand as the latter approach adds complexity you'd prefer the former still though?

Reply:
Fascinating stuff, thanks for sharing! I'm super curious what's consuming CPU there as I really wouldn't expect a 100ms delay to cause any increase. Can you try increasing it to a second just as a sanity check to see if CPU reduces further?

As an aside, another issue with the wait handle is that it's a blocking call, so each paused receive buffer will be consuming a threadpool thread while paused, preventing other work from being done. I wonder if that could be a cause of overall lower CPU? Is there much else going on on the machine?

Reply:
Here's what it looks like with 1000ms (green):

![image](https://github.com/justeattakeaway/JustSaying/assets/92371891/0cc7165f-e351-4d7d-b680-84092054fd58)

The orange is the active service process, as it toggles to inactive around 4:30pm the CPU increases (this is the 100ms version). Around 4:40pm I manually toggle so orange is now active, and green CPU doesn't increase - so it appears the longer delay does reduce CPU usage.

There are a couple other .NET services on the machine (a separate active/inactive application for each service), not much traffic. I think in our use case it doesn't matter some threads are blocked, as the application won't be serving traffic, but perhaps that's not the case for other users, and they want to pause listening for messages while doing other work - so from that perspective it makes sense to not make it blocking.

Reply:
Ok thanks, interesting. A quick benchmark of `ThrowIfCancellationRequested` doesn't suggest it should be an issue at all - each call is under 1ns even when using linked token sources.

Would you perhaps be able to share the code you're using to test the Task.Delay approach?

Thanks

Reply:
Ah - if it's the same approach as you shared earlier, no worries. I suspect the wait+continue approach would be simpler/easier to maintain even if it is a little slower (maybe a profiler could be attached to find the culprit).

Reply:
Sure, I can revert to the busy wait approach and make the other relevant changes Martin mentioned. I think it would be nice to be able to set the delay amount via a config setting just in case. Also would you be against having the extra while loop? It makes including logging easier so it only logs once on status change e.g. https://github.com/justeattakeaway/JustSaying/pull/1164/commits/2d341a3853eddafa439e4190af64a651f6eccb9f - alternatively the logging could be in the MessageReceiveController/Signaller

Reply:
Nice one, I think this is pretty straightforward now 🙂